[WIP] Initial alternative plugin system

KrauseFx opened the pull request:
Alternative to https://github.com/fastlane/fastlane/pull/4619

KrauseFx committed:
Initial alternative plugin system

KrauseFx commented:
I haven't yet looked into how we can actually load and use those gems then. 
I was assuming it would be nice to have separate lock files, but thinking about it, that's not what lock files are built-for.

I'll iterate once again, and use only one lock file. We can still have separate Gemfiles, so that it's easy for us to provide a CLI to add and remove plugins. 
All we have to do is add this to the `./Gemfile`:

```ruby
content = File.read(File.join(File.dirname(__FILE__), 'fastlane', 'FastlaneGemfile'))
eval(content, binding)
```

this way we still have separate Gemfiles, one in the `fastlane` directory, but only have one lock file.

This technique however doesn't allow us to update just the plugins without actually updating all the other things, right?

KrauseFx committed:
Removed separate lock files

KrauseFx committed:
Switch to Bundler and separate Gemfile

KrauseFx committed:
Some clean ups

KrauseFx committed:
Added xcversion plugin

KrauseFx committed:
Renamed gem

KrauseFx committed:
Added use of actual parameter for add_plugin

Commit:
Add a version range to bundler dependency

Bundler.setup in the trigger subcommand

KrauseFx committed:
Exec out to show password prompt

KrauseFx committed:
Added new PluginsLoader to automatically require all the things

KrauseFx committed:
Removed requires in the Fastfile, since it now works for free

KrauseFx committed:
Removed deprecated .all call

KrauseFx committed:
Added verbose debugging output

Commit:
Add a bit more support for creating a Gemfile from scratch

KrauseFx commented on fastlane/lib/fastlane/commands_generator.rb:
```diff
@@ -181,8 +183,16 @@ def run
 
         c.action do |args, options|
           plugin_name = args.last # TODO: actually parse the things
+          plugin_name = 'fastlane_' + plugin_name unless plugin_name.start_with?('fastlane_')
```
Maybe this should actually be part of the `PluginManager`

mfurtak replied:
Yup, seems right

KrauseFx commented on fastlane/lib/fastlane/plugin_manager.rb:
```diff
@@ -2,35 +2,43 @@ module Fastlane
         content = pluginsfile_content || "# Autogenerated by fastlane\n\n"
-        content += "gem '#{plugin_name}'\n"
+        # TODO: obviously need to stop hardcoding a git URL!
+        content += "gem '#{plugin_name}', git: 'git@github.com:fastlane/#{plugin_name}.git'\n"
```
By default, we'll just have `gem "fastlane_name"`, right? The git url should only be used when the owner doesn't push their gem to RubyGems

KrauseFx replied:
And during development this will be a local path

mfurtak replied:
Yes, as the comment says, the git business should not be there in the long term!

KrauseFx commented on fastlane/lib/fastlane/plugin_manager.rb:
```diff
@@ -40,56 +48,65 @@ def add_dependency(plugin_name)
       return true
     end
 
+    def create_gemfile(path)
```
Maybe add a `!` operator, to be consistent with `attach_plugins!`, both methods change local files

KrauseFx committed:
Moved loading of plugins to a later point to support --verbose

KrauseFx committed:
Added proper error message when plugin action couldn't be loaded

Commit:
Solve the mystery of failing bundle install

KrauseFx commented on fastlane/bin/fastlane:
```diff
@@ -1,5 +1,13 @@
 #!/usr/bin/env ruby
 $LOAD_PATH.push File.expand_path("../../lib", __FILE__)
 
+require 'fastlane_core'
```
We have to make sure to not forget the rocket emoji 🚀

KrauseFx replied:
Oh, you thought of it, never mind 👍

jcampbell05 commented:
Excited for this! Are you able to give an overview how this will work so I have some context when reviewing :)

KrauseFx committed:
Fixed code style errors

KrauseFx committed:
Added initial Plugins.md

KrauseFx committed:
Added note about how to test a plugin locally

KrauseFx committed:
Moved importing of actions into plugins

Related https://github.com/fastlane/fastlane_xcversion/commit/02899bcb0caa2a50d82bc2ff908cc25b9925238c

This is required due to allowing any number of actions in a plugin

KrauseFx committed:
Stop setting a static Git URL

KrauseFx committed:
Support for `fastlane add_plugin plugin1 plugin2`

KrauseFx committed:
Refactored the way lane switching and actions are called

This was a necessary refactor to make error handling cleaner when a plugin is not properly loaded. This will be a common issue and we need to offer appropriate error handling help

KrauseFx committed:
Introduced a constant for the prefix of fastlane plugins

KrauseFx committed:
Improved plugin detection and added initial test

mfurtak commented on fastlane/lib/fastlane/plugin_manager.rb:
```diff
@@ -24,8 +25,12 @@ def pluginsfile_content
+
     def add_dependency(plugin_name)
-      plugin_name = 'fastlane_' + plugin_name unless plugin_name.start_with?('fastlane_')
+      plugin_name = self.class.plugin_prefix + plugin_name unless plugin_name.start_with?(plugin_prefix)
```
`plugin_prefix` is being called two different ways on this line

KrauseFx commented:
Closing this PR to split up the work 👍
